src/main/java/at/hannibal2/skyhanni/features/chat/EmojiColorer.kt

```diff
@@ -0,0 +1,132 @@
+        "<('O')>" to "§r§e§l<('O')>",
+        "(ᵔᴥᵔ)" to "§r§6(ᵔᴥᵔ)",
+        "ヽ (◕◡◕) ﾉ" to "§r§aヽ (◕◡◕) ﾉ",
+        "= ＾● ⋏ ●＾ =" to "§r§e= §b＾● ⋏ ●＾§e =",
+        "☃" to "§r§b☃",
+        "(✿◠‿◠)" to "§r§e(§a✿§e◠‿◠)§7",
+        "ヽ(^◇^*)/" to "§r§eヽ(^◇^*)/",
+        "ヽ(⌐■_■)ノ♬" to "§r§9ヽ§5(§d⌐§c■§6_§e■§b)§3ノ§9♬",
+        "(・⊝・)" to "§r§6(§8・§6⊝§8・§6)",
+        "^-^" to "§r§a^-^",
+        "<o/" to "§r§d<§eo§d/",
+        "^_^" to "§r§a^_^"
+    )
+
+    private val resettingCodes = arrayOf(
+        "§0",
+        "§1",
+        "§2",
+        "§3",
+        "§4",
+        "§5",
+        "§6",
+        "§7",
+        "§8",
+        "§9",
+        "§a",
+        "§b",
+        "§c",
+        "§d",
+        "§e",
+        "§f"
+    )
+
+    private val formattingCodes = arrayOf(
+        "§k",
+        "§l",
+        "§m",
+        "§n",
+        "§o",
+        "§r"
+    )
+
+    @SubscribeEvent
+    fun onChatReceive(event: LorenzChatEvent) {
+        if (config.colorEmoji && EmojiReplacer.inArray(event.message.removeColor(), emojis)) {
+            val newMessage = replace(event, emojis)
+            event.blockedReason = "Emoji"
+            LorenzUtils.chat(newMessage, false)
+        }
+    }
+
+    private fun replace (event: LorenzChatEvent, array: Array<Pair<String, String>>): String {
+        if (EmojiReplacer.inArray(event.message.removeColor(), array)) {
+            var oldMessage = event.message
+            array.forEach { (search, replace) ->
+                if (oldMessage.contains(search)) {
+                    val oldColor = findColor(oldMessage, search)
+                    oldMessage = oldMessage.replace(search, "$replace$oldColor")
+                }
+            }
+            return oldMessage
+        }
+        return event.message
+    }
+
+    private fun findColor(message: String, search: String): String {
+        val lastIndex = message.indexOf(search)
+        val firstIndex = findFirstIndex(message, lastIndex)
+        val subString = message.substring(firstIndex,lastIndex)
+        return handleColor(subString)
+    }
+
+    private fun handleColor(subString: String): String {
+        var temp = false
+        var newSubstring = subString
+        resettingCodes.forEach { search ->
+            if (subString.contains(search) && search != "§r") temp = true
+        }
+        if (!temp) newSubstring = "§f$newSubstring"
+        var finalColors = "§${newSubstring[1]}"
+        for (code in formattingCodes) {
+            if (newSubstring.contains(code)) {
+                finalColors += code
+            }
+        }
+        return finalColors
+    }
+
+    private fun findFirstIndex(message: String, lastIndex: Int): Int {
+        val subString = message.substring(0, lastIndex)
+        var maxIndex = 0
+
+        resettingCodes.forEach { code ->
+            val index = subString.lastIndexOf(code)
+            if (index > maxIndex) maxIndex = index
+        }
+        return maxIndex
+    }
```

Comment:
This whole color logic feels off/overly complicated to me:
```suggestion
    /** Find a string that fully restores the current color state when prepended to another string */
    private fun findColorState(string: String): String {
        val lastColorIndex = resettingCodes.maxOf { string.lastIndexOf(it) }
        val lastColor = if (lastColorIndex < 0) "§r" else string.substring(lastColorIndex, lastColorIndex + 1)
        val restOfText = string.substring(it + 1)
        val modifiers = formattingCodes.filter { it in restOfText }
        return lastColor + modifiers.joinToString("")
    }
```

Keep in mind this requires `"§3"` to be deleted from `formattingCodes`, and instead be added to `resettingCodes`